File: src/main/java/me/aov/sellgui/gui/SellMenuConfig.java
package me.aov.sellgui.gui;

import me.aov.sellgui.SellGUIMain;
import me.aov.sellgui.utils.ItemIdentifier;
import org.bukkit.Material;
import org.bukkit.configuration.ConfigurationSection;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.inventory.ItemStack;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

public final class SellMenuConfig {
    public static final String DEFAULT_MENU_ID = "default";

    private final SellGUIMain plugin;
    private final String id;
    private final ConfigurationSection section;
    private final Set<String> allowedItems;
    private final Set<String> deniedItems;
    private final boolean exclusiveItems;

    private SellMenuConfig(SellGUIMain plugin, String id, ConfigurationSection section) {
        this.plugin = plugin;
        this.id = normalizeMenuId(id);
        this.section = section;
        this.allowedItems = loadItemSet(section, "item-filter.allowed-items", "allowed-items");
        this.deniedItems = loadItemSet(section, "item-filter.denied-items", "denied-items", "item-filter.blocked-items", "blocked-items");
        this.exclusiveItems = section.getBoolean("item-filter.exclusive", section.getBoolean("exclusive-items", false));
    }

    public static SellMenuConfig load(SellGUIMain plugin, String requestedId) {
        FileConfiguration guiConfig = plugin.getConfigManager().getGUIConfig();
        if (guiConfig == null) {
            return null;
        }

        String id = normalizeMenuId(requestedId);
        ConfigurationSection section = guiConfig.getConfigurationSection("sell_menus." + id);
        if (section == null && DEFAULT_MENU_ID.equals(id)) {
            section = guiConfig.getConfigurationSection("sell_gui");
        }

        return section == null ? null : new SellMenuConfig(plugin, id, section);
    }

    public static List<String> getMenuIds(SellGUIMain plugin) {
        FileConfiguration guiConfig = plugin.getConfigManager().getGUIConfig();
        if (guiConfig == null) {
            return Collections.singletonList(DEFAULT_MENU_ID);
        }

        LinkedHashSet<String> ids = new LinkedHashSet<>();
        ConfigurationSection menus = guiConfig.getConfigurationSection("sell_menus");
        if (menus != null) {
            for (String key : menus.getKeys(false)) {
                if (menus.isConfigurationSection(key)) {
                    ids.add(normalizeMenuId(key));
                }
            }
        }

        if (ids.isEmpty() && guiConfig.isConfigurationSection("sell_gui")) {
            ids.add(DEFAULT_MENU_ID);
        }

        if (ids.isEmpty()) {
            ids.add(DEFAULT_MENU_ID);
        }

        return new ArrayList<>(ids);
    }

    public static boolean menuExists(SellGUIMain plugin, String requestedId) {
        return load(plugin, requestedId) != null;
    }

    public static String normalizeMenuId(String rawId) {
        if (rawId == null || rawId.trim().isEmpty()) {
            return DEFAULT_MENU_ID;
        }
        return rawId.trim().toLowerCase(Locale.ROOT);
    }

    public static String normalizeItemKey(String raw) {
        if (raw == null || raw.trim().isEmpty()) {
            return null;
        }

        String key = raw.trim().toUpperCase(Locale.ROOT);
        if (!key.contains(":")) {
            key = "VANILLA:" + key;
        }
        return key;
    }

    public static boolean isExclusiveToAnyMenu(SellGUIMain plugin, ItemStack item) {
        for (String menuId : getMenuIds(plugin)) {
            SellMenuConfig menuConfig = load(plugin, menuId);
            if (menuConfig != null && menuConfig.isExclusiveItems() && menuConfig.matchesAllowedItem(item)) {
                return true;
            }
        }
        return false;
    }

    public boolean allowsItem(ItemStack item) {
        if (item == null || item.getType() == Material.AIR) {
            return false;
        }

        if (matchesDeniedItem(item)) {
            return false;
        }

        if (!allowedItems.isEmpty()) {
            return matchesAllowedItem(item);
        }

        return !isExclusiveToAnotherMenu(item);
    }

    private boolean isExclusiveToAnotherMenu(ItemStack item) {
        for (String menuId : getMenuIds(plugin)) {
            if (id.equals(menuId)) {
                continue;
            }

            SellMenuConfig menuConfig = load(plugin, menuId);
            if (menuConfig != null && menuConfig.isExclusiveItems() && menuConfig.matchesAllowedItem(item)) {
                return true;
            }
        }
        return false;
    }

    public boolean matchesAllowedItem(ItemStack item) {
        return matchesItemSet(item, allowedItems);
    }

    public boolean matchesDeniedItem(ItemStack item) {
        return matchesItemSet(item, deniedItems);
    }

    private boolean matchesItemSet(ItemStack item, Set<String> itemSet) {
        if (item == null || item.getType() == Material.AIR || itemSet.isEmpty()) {
            return false;
        }

        String identifier = normalizeItemKey(ItemIdentifier.getItemIdentifier(item));
        String material = normalizeItemKey(item.getType().name());
        return (identifier != null && itemSet.contains(identifier)) || itemSet.contains(material);
    }

    private static Set<String> loadItemSet(ConfigurationSection section, String... paths) {
        Set<String> items = new HashSet<>();
        for (String path : paths) {
            for (String raw : section.getStringList(path)) {
                String normalized = normalizeItemKey(raw);
                if (normalized != null) {
                    items.add(normalized);
                }
            }
        }
        return items;
    }

    public String getId() {
        return id;
    }

    public String getDisplayName() {
        return section.getString("name", id);
    }

    public String getPermission() {
        return section.getString("permission", "");
    }

    public String getString(String path, String fallback) {
        return section.getString(path, fallback);
    }

    public int getInt(String path, int fallback) {
        return section.getInt(path, fallback);
    }

    public boolean getBoolean(String path, boolean fallback) {
        return section.getBoolean(path, fallback);
    }

    public boolean contains(String path) {
        return section.contains(path);
    }

    public List<Integer> getIntegerList(String path) {
        return section.getIntegerList(path);
    }

    public List<String> getStringList(String path) {
        return section.getStringList(path);
    }

    public boolean isExclusiveItems() {
        return exclusiveItems;
    }
}

File: src/main/kotlin/me/aov/sellgui/gui/SellMenuConfig.kt
package me.aov.sellgui.gui

import me.aov.sellgui.SellGUIMain
import me.aov.sellgui.utils.ItemIdentifier
import org.bukkit.Material
import org.bukkit.configuration.ConfigurationSection
import org.bukkit.inventory.ItemStack
import java.util.Locale

class SellMenuConfig private constructor(private val plugin: SellGUIMain, private val id: String, private val section: ConfigurationSection) {
    private val allowedItems = loadItemSet(section, "item-filter.allowed-items", "allowed-items")
    private val deniedItems = loadItemSet(section, "item-filter.denied-items", "denied-items", "item-filter.blocked-items", "blocked-items")
    private val exclusiveItems = section.getBoolean("item-filter.exclusive", section.getBoolean("exclusive-items", false))
    fun allowsItem(item: ItemStack?): Boolean = item != null && item.type != Material.AIR && !matchesDeniedItem(item) && (allowedItems.isEmpty() && !isExclusiveToAnotherMenu(item) || allowedItems.isNotEmpty() && matchesAllowedItem(item))
    private fun isExclusiveToAnotherMenu(item: ItemStack): Boolean = getMenuIds(plugin).any { menuId -> menuId != id && load(plugin, menuId)?.let { it.isExclusiveItems() && it.matchesAllowedItem(item) } == true }
    fun matchesAllowedItem(item: ItemStack?): Boolean = matchesItemSet(item, allowedItems)
    fun matchesDeniedItem(item: ItemStack?): Boolean = matchesItemSet(item, deniedItems)
    private fun matchesItemSet(item: ItemStack?, itemSet: Set<String>): Boolean { if (item == null || item.type == Material.AIR || itemSet.isEmpty()) return false; return normalizeItemKey(ItemIdentifier.getItemIdentifier(item))?.let(itemSet::contains) == true || itemSet.contains(normalizeItemKey(item.type.name)) }
    fun getId(): String = id; fun getDisplayName(): String = section.getString("name", id) ?: id; fun getPermission(): String = section.getString("permission", "") ?: ""; fun getString(path: String, fallback: String): String = section.getString(path, fallback) ?: fallback; fun getInt(path: String, fallback: Int): Int = section.getInt(path, fallback); fun getBoolean(path: String, fallback: Boolean): Boolean = section.getBoolean(path, fallback); fun contains(path: String): Boolean = section.contains(path); fun getIntegerList(path: String): List<Int> = section.getIntegerList(path); fun getStringList(path: String): List<String> = section.getStringList(path); fun isExclusiveItems(): Boolean = exclusiveItems
    companion object {
        @JvmField val DEFAULT_MENU_ID = "default"
        @JvmStatic fun load(plugin: SellGUIMain, requestedId: String?): SellMenuConfig? { val config = plugin.configManager.guiConfig ?: return null; val id = normalizeMenuId(requestedId); val section = config.getConfigurationSection("sell_menus.$id") ?: if (id == DEFAULT_MENU_ID) config.getConfigurationSection("sell_gui") else null; return section?.let { SellMenuConfig(plugin, id, it) } }
        @JvmStatic fun getMenuIds(plugin: SellGUIMain): List<String> { val config = plugin.configManager.guiConfig ?: return listOf(DEFAULT_MENU_ID); val ids = linkedSetOf<String>(); config.getConfigurationSection("sell_menus")?.getKeys(false)?.filter { config.isConfigurationSection("sell_menus.$it") }?.forEach { ids += normalizeMenuId(it) }; if (ids.isEmpty() && config.isConfigurationSection("sell_gui")) ids += DEFAULT_MENU_ID; return if (ids.isEmpty()) listOf(DEFAULT_MENU_ID) else ids.toList() }
        @JvmStatic fun menuExists(plugin: SellGUIMain, requestedId: String?): Boolean = load(plugin, requestedId) != null
        @JvmStatic fun normalizeMenuId(rawId: String?): String = rawId?.trim()?.takeIf { it.isNotEmpty() }?.lowercase(Locale.ROOT) ?: DEFAULT_MENU_ID
        @JvmStatic fun normalizeItemKey(raw: String?): String? = raw?.trim()?.takeIf { it.isNotEmpty() }?.uppercase(Locale.ROOT)?.let { if (it.contains(':')) it else "VANILLA:$it" }
        @JvmStatic fun isExclusiveToAnyMenu(plugin: SellGUIMain, item: ItemStack?): Boolean = getMenuIds(plugin).any { load(plugin, it)?.let { config -> config.isExclusiveItems() && config.matchesAllowedItem(item) } == true }
        private fun loadItemSet(section: ConfigurationSection, vararg paths: String): Set<String> = paths.flatMap { section.getStringList(it) }.mapNotNull(::normalizeItemKey).toSet()
    }
}
